File: clients/go/internal/cli/autotendcmd.go
package cli

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"daisugi-verify/internal/capture"
	"daisugi-verify/internal/config"
	"daisugi-verify/internal/pathways"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
	"daisugi-verify/internal/tracejournal"
	"daisugi-verify/internal/verify"
)

func (e *Env) hook(args []string) error {
	if len(args) == 0 || args[0] == "--help" {
		e.out("Usage: daisugi hook [OPTIONS] COMMAND [ARGS]...\n\n  Capture hooks. This binary carries auto-tend.\n\n" +
			"Commands:\n  auto-tend  Close the captures to traces to distillation loop in one call.\n")
		return nil
	}
	if args[0] == "auto-tend" {
		return e.autoTend(args[1:])
	}
	return e.notYet("daisugi hook " + args[0])
}

// conversion is one session this run converts, checked before anything
// is written.
type conversion struct {
	session capture.Session
	task    string
	env     *pyjson.Object
	plan    *pyjson.Object
	result  *pyjson.Object
	// skip is why captures_to_trace raised a ValueError the command
	// prints and skips; the session is not converted.
	skip string
}

// prepareConversion is captures_to_trace up to the journal write: the
// records read, the envelope inferred, the plan built and verified. A
// capture whose trace would carry a violation is refused: the binary does
// not word every violation's detail yet.
func prepareConversion(s capture.Session, decompose bool) (*conversion, error) {
	recs, err := capture.Records(s.Path)
	if err != nil {
		return nil, err
	}
	if len(recs) == 0 {
		return nil, nil
	}
	task := "captured session " + s.ID
	env, err := capture.InferEnvelope(recs, task, decompose)
	if err != nil {
		return nil, err
	}
	plan, err := capture.Plan(recs, task)
	var invalid *capture.InvalidPlan
	if errors.As(err, &invalid) {
		return &conversion{session: s, skip: invalid.Text}, nil
	}
	if err != nil {
		return nil, err
	}
	p, err := verify.ParsePlan([]byte(pathways.DumpJSON(plan)))
	if err != nil {
		return nil, fmt.Errorf("%w: %v", capture.ErrUnreadable, err)
	}
	en, err := verify.ParseEnvelope([]byte(pathways.DumpJSON(env)))
	if err != nil {
		return nil, fmt.Errorf("%w: %v", capture.ErrUnreadable, err)
	}
	t0 := time.Now()
	res := verify.Verify(p, en, verify.VerifyOptions{Z3TimeoutMs: 500})
	ms := float64(time.Since(t0).Nanoseconds()) / 1e6
	if !res.OK {
		return nil, fmt.Errorf("%w: session %s would be journaled with a violation (%s), whose detail this binary does not write yet",
			capture.ErrUnreadable, s.ID, res.Violations[0].Message)
	}
	warnings := []any{}
	for _, w := range res.Timeouts {
		warnings = append(warnings, w)
	}
	result := pyjson.NewObject().Set("ok", true).Set("violations", []any{}).Set("warnings", warnings).
		Set("envelope_id", env.Value("id")).Set("plan_id", plan.Value("id")).Set("duration_ms", ms).
		Set("client", "python").Set("fallback", nil).Set("client_verdict", nil)
	// The body is written after other sessions are: one this binary
	// cannot write must refuse now, before the first write.
	if _, err := tracejournal.TraceBody(task, env, plan, result, "2000-01-01-00000000", "2000-01-01T00:00:00Z"); err != nil {
		return nil, err
	}
	return &conversion{session: s, task: task, env: env, plan: plan, result: result}, nil
}

func (e *Env) autoTend(args []string) error {
	const cmd = "hook auto-tend"
	opts := []opt{
		{names: []string{"--captures-root"}, value: true, metavar: "PATH"},
		{names: []string{"--data-dir"}, value: true, metavar: "PATH"},
		{names: []string{"--min-interval"}, value: true, metavar: "INTEGER", help: "Skip if last auto-tend was newer than this many seconds."},
		{names: []string{"--force"}, help: "Ignore the min-interval gate."},
		{names: []string{"--skip-distill"}, help: "Convert captures-to-traces but don't run tend afterwards."},
		{names: []string{"--allow-shell-decomposition"}, neg: "--no-allow-shell-decomposition",
			help: "Let the envelope admit compound shell (ADR-0010)."},
	}
	p, err := parseArgs(args, opts, 0)
	if err != nil {
		return e.usage(cmd, err)
	}
	if p.help {
		return e.cmdHelp(cmd, "", "Close the captures->traces->distillation loop in one cron-friendly call.", opts)
	}
	minInterval, err := clickInt(p, "--min-interval", 3600)
	if err != nil {
		return e.usage(cmd, err)
	}
	root := p.str("--captures-root", filepath.Join(e.home, ".opendaisugi", "captures"))
	dataDir := p.str("--data-dir", filepath.Join(e.home, ".opendaisugi"))
	force := p.flag("--force")
	cfg, err := config.Load(filepath.Join(dataDir, "config.yaml"))
	if errors.Is(err, config.ErrInvalid) {
		e.errf("daisugi %s: pydantic_core._pydantic_core.ValidationError: the config file does not validate\n", cmd)
		return exit(1)
	}
	if err != nil {
		return e.refuse(cmd, fmt.Errorf("the config file is not one this binary reads: %w", err))
	}
	if !force && (cfg.AutoTend == nil || !*cfg.AutoTend) {
		e.out("skipped: background distillation is off. Run `daisugi install` to opt in, or pass --force to tend once anyway.\n")
		return nil
	}
	stamp := filepath.Join(dataDir, ".hook-auto-tend-last-run")
	now := nowSeconds()
	last, err := readStamp(stamp)
	if err != nil {
		return e.failPy(cmd, err)
	}
	if !force && now-last < float64(minInterval) {
		e.out("skipped: last run %ds ago (< --min-interval=%d); use --force to override\n", int64(now-last), minInterval)
		return nil
	}
	decompose := cfg.ShellAllowDecomposition
	if p.flagSet("--allow-shell-decomposition") {
		decompose = p.flag("--allow-shell-decomposition")
	}
	// Everything this run would convert is read and checked first, so a
	// capture this binary cannot convert changes nothing.
	sessions, err := capture.ListSessions(root)
	if err != nil {
		return e.autoTendErr(cmd, err)
	}
	// Read only: the journal is made or migrated after every refusal.
	var jr *tracejournal.Journal
	if _, err := os.Stat(filepath.Join(dataDir, "journal", "index.db")); err == nil {
		if jr, err = tracejournal.OpenReadOnly(dataDir); err != nil {
			return e.failPy(cmd, err)
		}
		defer jr.Close()
	}
	var todo []*conversion
	var skipped []string
	for _, s := range sessions {
		if jr != nil {
			done, err := jr.IsConverted(s.ID)
			if err != nil {
				return e.failPy(cmd, err)
			}
			if done {
				continue
			}
		}
		c, err := prepareConversion(s, decompose)
		if err != nil {
			return e.autoTendErr(cmd, err)
		}
		if c == nil {
			skipped = append(skipped, s.ID)
		}
		todo = append(todo, c)
	}
	nconv := 0
	for _, c := range todo {
		if c != nil && c.skip == "" {
			nconv++
		}
	}
	if nconv > 0 && !p.flag("--skip-distill") {
		// The tend that follows the conversions must not refuse after
		// they are written: its checks run now, with the conversions
		// counted in.
		if _, err := e.tendCheck(dataDir, "anthropic/claude-sonnet-4-20250514", 3, 30, false, nconv, nowSeconds()); err != nil {
			// Daisugi() raises only after the conversions, in Python; that
			// one is reported when tend runs.
			var raised *raisedError
			if !errors.As(err, &raised) {
				return err
			}
		}
	}
	// The trace ids are drawn before the first write, so a failed draw
	// changes nothing.
	ids := map[string]string{}
	for _, c := range todo {
		if c != nil && c.skip == "" {
			if ids[c.session.ID], err = randHex8(); err != nil {
				return e.failPy(cmd, err)
			}
		}
	}
	if jr != nil {
		jr.Close()
	}
	j, err := tracejournal.Open(dataDir)
	if err != nil {
		return e.failPy(cmd, err)
	}
	defer j.Close()
	var converted []string
	for _, s := range sessions {
		c := findConversion(todo, s.ID)
		if c != nil && c.skip != "" {
			e.errf("  skipped %s: %s\n", s.ID, c.skip)
			continue
		}
		if c == nil {
			for _, sk := range skipped {
				if sk == s.ID {
					e.errf("  skipped %s: no records in %s\n", s.ID, filepath.Join(root, s.ID+".jsonl"))
				}
			}
			continue
		}
		created := time.Now().UTC().Format("2006-01-02T15:04:05Z")
		traceID := created[:10] + "-" + ids[s.ID]
		if err := j.Log(c.task, c.env, c.plan, c.result, traceID, created); err != nil {
			return e.autoTendErr(cmd, err)
		}
		if err := j.MarkConverted(s.ID, traceID, nowSeconds()); err != nil {
			return e.failPy(cmd, err)
		}
		converted = append(converted, traceID)
		e.out("  converted %s → %s\n", s.ID, traceID)
	}
	e.out("converted %d sessions\n", len(converted))
	if len(converted) > 0 && !p.flag("--skip-distill") {
		rep, err := e.runTendQuiet(dataDir)
		if err != nil {
			return err
		}
		if rep != nil {
			e.out("tend: created=%d updated=%d skipped=%d\n", rep.Created, rep.Updated, rep.Skipped)
		}
	}
	if err := os.MkdirAll(dataDir, 0o777); err != nil {
		return e.failPy(cmd, err)
	}
	if err := os.WriteFile(stamp, []byte(pyjson.FloatRepr(now)), 0o666); err != nil {
		return e.failPy(cmd, err)
	}
	return nil
}

// runTendQuiet is Daisugi(data_dir=...).tend() as auto-tend runs it: the
// defaults, and a failure reported as "tend failed" rather than raised.
func (e *Env) runTendQuiet(dataDir string) (*tendReport, error) {
	e.tendFailed = true
	defer func() { e.tendFailed = false }()
	rep, err := e.runTend(dataDir, "anthropic/claude-sonnet-4-20250514", 3, 30, false)
	if err != nil {
		if e.raised {
			return nil, err
		}
		var x *exitError
		if errors.As(err, &x) && x.code == 1 {
			// runTend printed its reason as the command's error; auto-tend
			// catches the exception and goes on.
			return nil, nil
		}
		return nil, err
	}
	return &rep, nil
}

func findConversion(todo []*conversion, id string) *conversion {
	for _, c := range todo {
		if c != nil && c.session.ID == id {
			return c
		}
	}
	return nil
}

// randRead is crypto/rand.Read; a test replaces it.
var randRead = rand.Read

// randHex8 is secrets.token_hex(4). A failed read is an error, never a
// fixed id.
func randHex8() (string, error) {
	b := make([]byte, 4)
	if _, err := randRead(b); err != nil {
		return "", fmt.Errorf("no random trace id: %w", err)
	}
	return hex.EncodeToString(b), nil
}

func (e *Env) autoTendErr(cmd string, err error) error {
	if errors.Is(err, capture.ErrUnreadable) || errors.Is(err, tracejournal.ErrUnreadable) {
		return e.refuse(cmd, err)
	}
	return e.failPy(cmd, err)
}

// tendFailure is how auto-tend reports a tend that raised: "tend failed:
// <exception class name>: <message>".
func tendFailure(msg string) string {
	head, rest, found := strings.Cut(msg, ": ")
	if found && !strings.Contains(head, " ") {
		if i := strings.LastIndex(head, "."); i >= 0 {
			head = head[i+1:]
		}
		return "tend failed: " + head + ": " + rest
	}
	return "tend failed: " + msg
}

var _ = pystr.Repr

File: clients/rust/src/cli/autotendcmd.rs
//! `daisugi hook auto-tend`: captured sessions turned into journal traces,
//! then a tend, with the flags, output, exit codes and files of the Python
//! CLI. Every conversion is read and checked before the first write.

use std::cell::RefCell;
use std::rc::Rc;
use std::time::Instant;

use super::gardencmd::{now_seconds, read_stamp};
use super::tendcmd::{CheckStop, TendReport, DEFAULT_MODEL};
use super::{exit, parse_args, Env, Opt, Res, Stop};
use crate::capture::{infer_envelope, list_sessions, plan, records, Session, Skip};
use crate::distill::token_hex;
use crate::gate::pyjson::{float_repr, Object, Value};
use crate::pathways::PwErr;
use crate::shell_decompose::ShellParser;
use crate::tracejournal::{iso_seconds, trace_body, Journal};

/// One session this run converts, checked before anything is written.
struct Conversion {
    session: Session,
    task: String,
    env: Object,
    plan: Object,
    result: Object,
}

/// What a session comes to: a conversion, a skip with its text, or no
/// records at all.
enum Prepared {
    Convert(Box<Conversion>),
    Skip(Session, String),
    Empty(Session),
}

/// `captures_to_trace` up to the journal write: the records read, the
/// envelope inferred, the plan built and verified. A capture whose trace
/// would carry a violation is refused: the binary does not word every
/// violation's detail yet.
fn prepare_conversion(parser: &mut ShellParser, s: &Session, decompose: bool) -> Result<Prepared, PwErr> {
    let recs = records(&s.path)?;
    if recs.is_empty() {
        return Ok(Prepared::Empty(s.clone()));
    }
    let task = format!("captured session {}", s.id);
    let env = infer_envelope(parser, &recs, &task, decompose)?;
    let plan = match plan(&recs, &task)? {
        Ok(p) => p,
        Err(Skip::InvalidPlan(text)) => return Ok(Prepared::Skip(s.clone(), text)),
    };
    let t0 = Instant::now();
    let out = crate::pathways::verify::verify(&Value::Obj(plan.clone()), &Value::Obj(env.clone()), None, 500)?;
    let ms = t0.elapsed().as_secs_f64() * 1000.0;
    if let Some(v) = out.violations.first() {
        return Err(PwErr::Unreadable(format!(
            "session {} would be journaled with a violation ({}), whose detail this binary does not write yet",
            s.id, v.message
        )));
    }
    let mut result = Object::new();
    result.set("ok", true);
    result.set("violations", Value::List(vec![]));
    result.set("warnings", Value::List(out.timeouts.into_iter().map(Value::Str).collect()));
    result.set("envelope_id", env.value("id").clone());
    result.set("plan_id", plan.value("id").clone());
    result.set("duration_ms", ms);
    result.set("client", "python");
    result.set("fallback", Value::Null);
    result.set("client_verdict", Value::Null);
    // The body is written after other sessions are: one this binary cannot
    // write must refuse now, before the first write.
    trace_body(&task, &env, &plan, &result, "2000-01-01-00000000", "2000-01-01T00:00:00Z")?;
    Ok(Prepared::Convert(Box::new(Conversion { session: s.clone(), task, env, plan, result })))
}

impl Env {
    pub(super) fn hook(&mut self, args: &[String]) -> Res {
        if args.is_empty() || args[0] == "--help" {
            self.out(
                "Usage: daisugi hook [OPTIONS] COMMAND [ARGS]...\n\n  Capture hooks. This binary carries auto-tend.\n\n\
                 Commands:\n  auto-tend  Close the captures to traces to distillation loop in one call.\n",
            );
            return Ok(());
        }
        if args[0] == "auto-tend" {
            return self.auto_tend(&args[1..]);
        }
        let what = format!("daisugi hook {}", args[0]);
        self.not_yet(&what)
    }

    fn auto_tend(&mut self, args: &[String]) -> Res {
        const CMD: &str = "hook auto-tend";
        let opts = [
            Opt::val(&["--captures-root"], "PATH", "Where the captured sessions are."),
            Opt::val(&["--data-dir"], "PATH", "Daisugi data directory."),
            Opt::val(&["--min-interval"], "INTEGER", "Skip if last auto-tend was newer than this many seconds."),
            Opt::flag(&["--force"], "Ignore the min-interval gate."),
            Opt::flag(&["--skip-distill"], "Convert captures-to-traces but don't run tend afterwards."),
            Opt::pair(&["--allow-shell-decomposition"], "--no-allow-shell-decomposition", "Let the envelope admit compound shell (ADR-0010)."),
        ];
        let p = parse_args(args, &opts, 0).map_err(|m| self.usage_stop(CMD, m))?;
        if p.help {
            return self.cmd_help(CMD, "", "Close the captures->traces->distillation loop in one cron-friendly call.", &opts);
        }
        let min_interval = self.click_int(CMD, &p, "--min-interval", 3600)?;
        let root = p.str("--captures-root", &format!("{}/.opendaisugi/captures", self.home));
        let data_dir = p.str("--data-dir", &format!("{}/.opendaisugi", self.home));
        let force = p.flag("--force");
        let cfg = match super::config::load(&format!("{data_dir}/config.yaml")) {
            Ok(c) => c,
            Err(super::config::ConfigErr::Invalid) => {
                self.errf(&format!("daisugi {CMD}: pydantic_core._pydantic_core.ValidationError: the config file does not validate\n"));
                return exit(1);
            }
            Err(e) => return self.refuse(CMD, &format!("the config file is not one this binary reads: {e}")),
        };
        if !force && cfg.auto_tend != Some(true) {
            self.out("skipped: background distillation is off. Run `daisugi install` to opt in, or pass --force to tend once anyway.\n");
            return Ok(());
        }
        let stamp = format!("{data_dir}/.hook-auto-tend-last-run");
        let now = now_seconds();
        let last = read_stamp(&stamp).map_err(|e| self.pw_err(CMD, e))?;
        if !force && now - last < min_interval as f64 {
            self.out(&format!(
                "skipped: last run {}s ago (< --min-interval={min_interval}); use --force to override\n",
                (now - last) as i64
            ));
            return Ok(());
        }
        let decompose =
            if p.flag_set("--allow-shell-decomposition") { p.flag("--allow-shell-decomposition") } else { cfg.shell_allow_decomposition };
        // Everything this run would convert is read and checked first, so a
        // capture this binary cannot convert changes nothing.
        let sessions = list_sessions(&root).map_err(|e| self.pw_err(CMD, e))?;
        // Read only: the journal is made or migrated after every refusal.
        let jr = if std::fs::metadata(format!("{data_dir}/journal/index.db")).is_ok() {
            Some(Journal::open_read_only(&data_dir).map_err(|e| self.pw_err(CMD, e))?)
        } else {
            None
        };
        let mut parser = ShellParser::new();
        let mut todo: Vec<Prepared> = vec![];
        for s in &sessions {
            if let Some(j) = &jr {
                if j.is_converted(&s.id).map_err(|e| self.pw_err(CMD, e))? {
                    continue;
                }
            }
            let c = prepare_conversion(&mut parser, s, decompose).map_err(|e| self.pw_err(CMD, e))?;
            todo.push(c);
        }
        drop(jr);
        let nconv = todo.iter().filter(|c| matches!(c, Prepared::Convert(_))).count();
        if nconv > 0 && !p.flag("--skip-distill") {
            // The tend that follows the conversions must not refuse after
            // they are written: its checks run now, with the conversions
            // counted in. Daisugi() raises only after the conversions, in
            // Python; that one is reported when tend runs.
            let notes = Rc::new(RefCell::new(String::new()));
            match self.tend_check(&data_dir, DEFAULT_MODEL, 3, 30, false, nconv, now_seconds(), &notes) {
                Ok(_) | Err(CheckStop::Raised(_)) => {}
                Err(CheckStop::Stop(s)) => return Err(s),
            }
        }
        // The trace ids are drawn before the first write, so a failed draw
        // changes nothing.
        let mut ids: Vec<String> = vec![];
        for c in &todo {
            if matches!(c, Prepared::Convert(_)) {
                ids.push(token_hex().map_err(|e| self.pw_err(CMD, e))?);
            }
        }
        let j = Journal::open(&data_dir).map_err(|e| self.pw_err(CMD, e))?;
        let mut converted = vec![];
        let mut ids = ids.into_iter();
        for c in &todo {
            match c {
                Prepared::Skip(s, text) => self.errf(&format!("  skipped {}: {text}\n", s.id)),
                Prepared::Empty(s) => self.errf(&format!("  skipped {}: no records in {}\n", s.id, s.path)),
                Prepared::Convert(c) => {
                    let created = format!("{}Z", iso_seconds(now_seconds() as i64));
                    let trace_id = format!("{}-{}", &created[..10], ids.next().unwrap_or_default());
                    j.log(&c.task, &c.env, &c.plan, &c.result, &trace_id, &created).map_err(|e| self.pw_err(CMD, e))?;
                    j.mark_converted(&c.session.id, &trace_id, now_seconds()).map_err(|e| self.pw_err(CMD, e))?;
                    self.out(&format!("  converted {} \u{2192} {trace_id}\n", c.session.id));
                    converted.push(trace_id);
                }
            }
        }
        drop(j);
        self.out(&format!("converted {} sessions\n", converted.len()));
        if !converted.is_empty() && !p.flag("--skip-distill") {
            if let Some(r) = self.run_tend_quiet(&data_dir)? {
                self.out(&format!("tend: created={} updated={} skipped={}\n", r.rep.created, r.rep.updated, r.rep.skipped));
            }
        }
        let write = std::fs::create_dir_all(&data_dir).and_then(|_| std::fs::write(&stamp, float_repr(now)));
        if let Err(e) = write {
            return Err(self.pw_err(CMD, PwErr::Io(e)));
        }
        Ok(())
    }

    /// `Daisugi(data_dir=...).tend()` as auto-tend runs it: the defaults,
    /// and a failure reported as "tend failed" rather than raised.
    fn run_tend_quiet(&mut self, data_dir: &str) -> Result<Option<TendReport>, Stop> {
        self.tend_failed = true;
        let r = self.run_tend(data_dir, DEFAULT_MODEL, 3, 30, false);
        self.tend_failed = false;
        match r {
            Ok(r) => Ok(Some(r)),
            Err(s) if self.raised => Err(s),
            // run_tend printed its reason as the command's error; auto-tend
            // catches the exception and goes on.
            Err(Stop::Exit(1)) => Ok(None),
            Err(s) => Err(s),
        }
    }
}
